runner/runners/queue.go

```diff
@@ -85,36 +86,58 @@ func NewQueueRunner(
 	}
 
 	statusManager := NewStatusManager(history)
-	inv := NewInvoker(exec, filer, output, tmp, stat)
+	inv := NewInvoker(exec, filerMap, output, tmp, stat)
 
 	controller := &QueueController{
 		statusManager: statusManager,
 		inv:           inv,
-		filer:         filer,
+		filerMap:      filerMap,
+		updateReq:     make(map[runner.RunType]bool),
 		capacity:      capacity,
 		reqCh:         make(chan interface{}),
 		updateCh:      make(chan interface{}),
-		cancelTimerCh: make(chan interface{}, 1),
+		cancelTimerCh: make(chan interface{}, len(filerMap)),
 	}
 	run := &Service{controller, statusManager, statusManager}
 
-	// QueueRunner will not serve requests if an idc is defined and returns an error
-	log.Info("Starting goroutine to check for snapshot init? ", (idc != nil))
-	var err error = nil
-	if idc != nil {
+	// QueueRunner waits on filers with InitDoneChannels defined to return,
+	// and will not serve requests if any return an error
+	var wg sync.WaitGroup
+	var initErr error = nil
+	wait := false
+
+	for t, f := range filerMap {
+		if f.IDC != nil {
+			log.Infof("Starting goroutine to wait on init for filer %v", t)
+			wg.Add(1)
+			wait = true
+
+			go func(rt runner.RunType, idc snapshot.InitDoneCh) {
+				err := <-idc
+				if err != nil {
+					initErr = err
+					log.Errorf("Init channel for filer %v returned error: %s", rt, err)
+				}
+				wg.Done()
+			}(t, f.IDC)
+		}
+	}
+
+	if wait {
+		// Wait for initialization to complete in a new goroutine to let this constructor return
 		go func() {
-			err = <-idc
-			if err != nil {
+			wg.Wait()
+			if initErr != nil {
```

Comment:
initErr only gets set if the IDC returned a non-nil error, so although it can get overwritten, it can't get overwritten with nil (and we log all the individual failures)

Reply:
ah yep you're right - missed that